### PR TITLE
Js traderefactor

### PR DESCRIPTION
**Pull request by jacobstringer**
This PR gives:

- A complete rewrite of trade functionality
- Updates of trade tests
- Bugfix where trade was not giving neighbours money due to DB committing problems
- Guard against null pointer on the game setup screen as suggested by Mitchell
- Create user has profile added to the return. @Marthijn-B is this what you wanted?

Next steps:
- With a couple of extra pieces of code, the trade function can replace the check valid move function, allowing for a big simplification of turn processing logic
- Once we are sure everything is working as it should (I did quite a bit of testing so it should be good), we can delete the old trade code and move checking code
- Since it doesn't take too much computation, we could add a summary of whether a card is playable and, if so, how much trade is needed, to each card in a hand sent to the front end. This would allow immediate and detailed feedback for cards, but I don't know how easy or hard this would be for the front end as opposed to the alternative, where the front end calls a new API for that information. Something to discuss for tonight's meeting @Marthijn-B, @wackydapper, @mmdonaldson?

### CODE REVIEW

File: application/app.py [hunk 1]
```diff
@@ -268,7 +269,7 @@ def play_card():
     from_discard_pile = false_true(request.args.get('from_discard_pile'))
     trade = false_true(request.args.get('trade'))
 
-    if process_card(card, player, discarded, for_wonder, from_discard_pile=from_discard_pile, trade=trade):
+    if process_card(card, player, discarded, for_wonder, from_discard_pile=from_discard_pile, tr=trade):
```
Comment (mmdonaldson) [inline application/app.py:272]: Hmm changing trade to tr is a bit implicit. I'd say should_trade is a bit more obvious and easier to read

File: application/controllers/card_logic.py [hunk 2]
```diff
@@ -69,6 +69,178 @@ def process_card(card, player, is_discarded, for_wonder, from_discard_pile=None,
...
+                temp = card_options[index]
+                for i in temp.keys():
+                    if i != 'Alternating' and temp[i][1] > current_price:
+                        c = card_options.pop(0)
+                        current_price = c[i][1]
+                        break
+                    elif i != 'Alternating' and temp[i][1] == current_price:
+                        c = card_options.pop(index)
+                        break
+                    else:
```
Comment (mmdonaldson) [inline application/controllers/card_logic.py:123]: Is there never another possibility? We should probably have something to catch a potential future error?

File: static/src/components/GameScreen.js [hunk 3]
```diff
@@ -222,8 +222,10 @@ export class GameScreen extends Component {
         if (!started) {
             this.playersLogged();
         }
-        if (started && game.game.age) {
-            document.title = `Age: ${game.game.age} Round: ${game.game.round}`;
+        if (started && game) {
+            if (game.game.age) {
```
Comment (mmdonaldson) [inline static/src/components/GameScreen.js]: love it

**Review by mmdonaldson — CHANGES_REQUESTED** [review]

File: static/src/components/GameScreen.js (hunk 3 above)
Comment (jacobstringer) [reply static/src/components/GameScreen.js]: That's elegant coding right there!

File: application/controllers/card_logic.py (hunk 2 above)
Comment (jacobstringer) [reply application/controllers/card_logic.py:123]: There's never any other possibilities afaik, as the 'cards' get created by get_card_benefits, which always gives an Alternating key with a boolean, and then everything else is a key with an array.

def get_card_benefits(card, prices, player):
    temp = {}
    if card.giveStone:
        temp['Stone'] = [card.giveStone, prices[0], player]
    if card.giveBrick:
        temp['Brick'] = [card.giveBrick, prices[0], player]
    if card.giveOre:
        temp['Ore'] = [card.giveOre, prices[0], player]
    if card.giveWood:
        temp['Wood'] = [card.giveWood, prices[0], player]
    if card.giveGlass:
        temp['Glass'] = [card.giveGlass, prices[1], player]
    if card.givePaper:
        temp['Paper'] = [card.givePaper, prices[1], player]
    if card.giveCloth:
        temp['Cloth'] = [card.giveCloth, prices[1], player]
    if card.resourceAlternating:
        temp['Alternating'] = True
    else:
        temp['Alternating'] = False
    return temp

File: application/app.py (hunk 1 above)
Comment (jacobstringer) [reply application/app.py:272]: Okay, will change that on the next PR I push (today).

**Comment by jacobstringer** [comment]
I'll close this and create a PR for the more comprehensive version I've been working on today.

**Pull request closed without merging**